main/lsp/LSPPreprocessor.h

```diff
@@ -0,0 +1,115 @@
+    // standard format.
+    UnorderedSet<std::string> openFiles;
+
+    // ErrorQueue used for finalGS. Separate from initialGS's ErrorQueue to prevent indexing errors from making it to
+    // the client.
+    std::shared_ptr<core::ErrorQueue> finalGSErrorQueue;
+
+    // Indicates the next ID to use on an incoming message. Used to refer to messages by ID.
+    int nextMessageId = 1;
+
+    /**
+     * Merges all consecutive file updates into a single update. File updates are also merged if they are only separated
+     * by *delayable* requests (see LSPMessage::isDelayable()). Updates are merged into the earliest file update in the
+     * sequence.
+     *
+     * Example: (E = edit, D = delayable non-edit, M = arbitrary non-edit)
+     * {[M1][E1][E2][D1][E3]} => {[M1][E1-3][D1]}
+     */
+    void mergeFileChanges(QueueState &state);
+
+    std::unique_ptr<LSPMessage> makeAndCommitWorkspaceEdit(std::unique_ptr<SorbetWorkspaceEditParams> params,
+                                                           std::unique_ptr<SorbetWorkspaceEditCounts> counts,
+                                                           std::unique_ptr<LSPMessage> oldMsg);
+
+    /* The following methods convert edits into LSPFileUpdates. */
+
+    void canonicalizeEdits(std::unique_ptr<DidChangeTextDocumentParams> changeParams, LSPFileUpdates &updates) const;
+    void canonicalizeEdits(std::unique_ptr<DidOpenTextDocumentParams> openParams, LSPFileUpdates &updates) const;
+    void canonicalizeEdits(std::unique_ptr<DidCloseTextDocumentParams> closeParams, LSPFileUpdates &updates) const;
+    void canonicalizeEdits(std::unique_ptr<WatchmanQueryResponse> queryResponse, LSPFileUpdates &updates) const;
+    void mergeEdits(int toId, LSPFileUpdates &to, int fromId, LSPFileUpdates &from);
+
+    /**
+     * Returns a global state for typechecking, cloned from initialGS. Note: The clone does not share an error queue
+     * with initialGS.
+     */
+    std::unique_ptr<core::GlobalState> getTypecheckingGS() const;
+
+public:
+    LSPPreprocessor(std::unique_ptr<core::GlobalState> initialGS, LSPConfiguration config, WorkerPool &workers,
+                    const std::shared_ptr<spdlog::logger> &logger);
+
+    /**
+     * Performs pre-processing on the incoming LSP request and appends it to the queue.
+     *
+     * * Merges changes to the same document + Watchman filesystem updates.
+     * * Processes pause/ignore requests.
+     * * Updates `openFiles` if a file opens/closes.
+     * * Updates `config` if the client sends an `initialize` message.
+     * * Indexes all files on filesystem if client sends an `initialized` message. If configured, will also send
+     * progress notifications.
+     *
+     * If a mutex is specified, it grabs the mutex before reading/writing state.
+     */
+    void preprocessAndEnqueue(QueueState &state, std::unique_ptr<LSPMessage> msg,
+                              std::optional<absl::Mutex *> stateMtx = std::nullopt);
```

Comment:
is a non-null `stateMtx` actually ever passed?

Reply:
also, if we'll need the mutext this seems "potentially unsafe by default", can we make it "conservative by default"?

Reply:
> is a non-null stateMtx actually ever passed?

Yes, for the single-threaded case (see `processMessage(unique_ptr<LSPMessage>)` in `LSPLoop`).

> can we make it "conservative by default"

What do you mean?

Reply:
>     can we make it "conservative by default"

make it lock by default rather than not

Reply:
ACK, I have the change locally, will push once I've finished working through some tests.